src/effects/builtin/common/generatoreffectmodel.h

```diff
@@ -14,23 +14,33 @@ class GeneratorEffectModel : public AbstractEffectModel
     Q_OBJECT
     Q_PROPERTY(double duration READ duration WRITE setDuration NOTIFY durationChanged)
     Q_PROPERTY(QString durationFormat READ durationFormat WRITE setDurationFormat NOTIFY durationFormatChanged)
-    Q_PROPERTY(double sampleRate READ sampleRate NOTIFY sampleRateChanged)
+    Q_PROPERTY(double sampleRate READ sampleRate NOTIFY sampleRateChanged FINAL)
+    Q_PROPERTY(double tempo READ tempo NOTIFY tempoChanged FINAL)
+    Q_PROPERTY(int upperTimeSignature READ upperTimeSignature NOTIFY upperTimeSignatureChanged FINAL)
+    Q_PROPERTY(int lowerTimeSignature READ lowerTimeSignature NOTIFY lowerTimeSignatureChanged FINAL)
 
 public:
 
     double duration() const;
     void setDuration(double newDuration);
     double sampleRate() const;
+    double tempo() const;
+    int upperTimeSignature() const;
+    int lowerTimeSignature() const;
     QString durationFormat() const;
     void setDurationFormat(const QString& newDurationFormat);
 
 signals:
     void sampleRateChanged();
+    void tempoChanged();
+    void upperTimeSignatureChanged();
+    void lowerTimeSignatureChanged();
     void durationChanged();
     void durationFormatChanged();
 
 private:
-    void doReload() override;
+    void doReload() final override;
+    virtual void doEmitSignals() = 0;
```

Comment:
Could you explain why we need another method?

Reply:
what if we want to add some logic to doReload that is different from sending signals?
there is a high probability that we will add this, then we will revert your change?

Reply:
I don't understand your concern. If `GeneratorEffectModel::doReload()` needs to do yet something else, then it does something else, and still call `doEmitSignals()`.